### ISSUE TITLE
[QUESTION] Query Conditional v1

### ISSUE THREAD

**Issue by z-murphydev** [dynamoose/dynamoose#865]
### Summary:

So, currently, I am trying to query my dynamoDB table. I want to find the record where the id is equal to the provided id (one column) then once I have that record - look a the column associated with it (companies) which has a list of Objects, and parse through that list of object to find only objects in that list who have nested properties that match what I'm looking for.

I've read the documentation on this, and I'm having the following issues.

1. The documentation for Conditionals says 'The Condition object represents a conditional that you can attach to various settings in other Dynamoose methods.' but I'm not seeing anywhere that says what methods these conditions can be attached to.

2. When attempting to use new dynamoose.Condition(options) I receive and error that says dynamoose.Condition is not a constructor, if I remove the 'new' keyword, I'm told it's not a function.

3. The objects in the list that I am looking to get have no specific shape/values so they're not defined in-depth on the schema. Not sure if this matters since when I view the data through the dashboard it all looks fine, and if I just run a standard get I can read the data, but figured I'd include it.

#### Schema
```
const dynamoose = require('dynamoose');
const Schema = dynamoose.Schema;
const CompaniesSchema = new Schema(
  {
    id: {
      type: String,
      required: true,
      trim: true,
    },
    companies: {
      type: 'list',
      list: [Map],
      required: true,
    },
  },
  {
    useDocumentTypes: true, //don't store arrays as jsonstring
    saveUnknown: true,
    useNativeBooleans: true,
  }
);
export = dynamoose.model(process.env.DYNAMO_COMPANIES_TABLE, CompaniesSchema);
```

#### General
```
    const filter = new dynamoose.Condition('id').eq('testid-1234');

    const companies = await CompaniesModel.query(filter).exec();
```

Ideally, what I'd like to be able to do is the following:
```
const companies = await CompaniesModel.query('id').eq('testid-1234').and().('companies.properties.industry').eq('telcomm')
```
The above query should return all Companies associated with the ID that have the industry telcomm (and industry is nested 2 levels deep, in an array of other objects with a similar, but not exact same, structure.)
### Environment:

Operating System: Windows 10
Operating System Version: 10.0.18362
Node.js version (`node -v`): 12.16.0
NPM version: (`npm -v`): 6.13.4
Dynamoose version: 1.10.0

### Other information (if applicable):

While some other questions have been similar to this, the solutions which were recommend to them (including FilterExpressions) did not work for me. Though I could have been implementing them incorrectly.

### Other:
- [x ] I have read through the Dynamoose documentation before posting this issue
- [x ] I have searched through the GitHub issues (including closed issues) and pull requests to ensure this question has not already been raised before
- [x ] I have searched the internet and Stack Overflow to ensure this question hasn't been raised or answered before
- [x ] I have filled out all fields above
- [x ] I am running the latest version of Dynamoose

**Comment by fishcharlie** [comment]
@z-murphydev Few things.

1. It looks like `id` is your hashKey, and you don't have any rangeKey. This means that all `id` values in your table must be unique. If you already know your `id` you can use `CompaniesModel.get`. There isn't any need to filter or query anymore since each `id` MUST be unique.
2. Conditionals were released in v2 of Dynamoose. This removes the need for things like `useDocumentTypes` and `useNativeBooleans` and has some other great changes. It was a full rewrite from the ground up. So sadly, you are unable to use Conditionals in v1. Likely you were looking at the documentation for v2 at: https://dynamoosejs.com. You can find the documentation for v1 at: https://v1.dynamoosejs.com.

I think this addresses the major things I saw in your question. Please let me know if you have any further questions.

You can also find an incomplete list of breaking changes for v2 in the [release notes](https://github.com/dynamoose/dynamoose/releases/tag/v2.0.0).

Also, if you need help upgrading to v2 or run into any problems with migrating, **please** reach out. I would be more than happy to help. There is a link to a Slack channel as well that you can join and ping me with more specific questions to your setup. Or you can continue to create issues. Whatever you prefer!

**Comment by z-murphydev** [comment]
@fishcharlie 

Thanks for the info! The clarification around the documentation is awesome, so thank you!

I'm not sure that using get really works for my use-case, though. All id's are unique - so I can get the correct object, but I was/am hoping that 'companies' would be attached to the document similarly to Mongo/Mongoose so that I would be able to have that takes an id and gets the companies associated with that id, then only returns companies that match a specific criteria.

When I try to make a batchGet like so:
```
    const companies = await CompaniesModel.batchGet(
      [{ id: 'test-1234' }],
      { 'companies.properties.industry': 'telcomm' }
    );
```

It just returns the object, with all of the companies attached instead of just the ones where the industry is telcomm.

For reference, this is a trimmed version of what a company could look like:
```
{
comapny_id: '123456',
properties: {
  name: 'Company1',
  domain: 'company.com',
  industry: 'telcomm'
 }
}
```

Do I need to define the full structure of company objects in the schema in order to be able to perform look-ups on those keys? Or does DynamDB/Dynamoose just not operate as similarly to Mongo/Mongoose as I thought and I just need to go grab all the companies, and then filter them once they've been returned from the DB through logic on the server?

**Comment by fishcharlie** [comment]
@z-murphydev Ahhhh. I think I understand what you are trying to do. So you want to get the document where ID = _____. But then only have it return the companies in the array that match a given condition.

Sadly, I don't think there is any simple way to do this in DynamoDB the way you are thinking about it. I think you might have to restructure the way you are thinking about the data. Maybe by having each company be a separate document or something? Not quite sure your use case fully so hard to advise here.

The only other solution would be to retrieve the entire document, and just get the data that you need on your end. Downside to this is you are chewing through more bandwidth. Might not be an issue for your use case tho to use that extra bandwidth. Looks like you might have suggested this as well. Might be a good route to go down.

The most similar thing to what you are talking about is called [`attributes`](https://v1.dynamoosejs.com/api/query/#queryattributesattributes), which is basically a list of attributes you wish you return. Problem with that, is you can't do conditionals or logic with it to only get the attributes you want. It's just an array of strings that represent the attributes you want it to return in the response. You can't do any conditional logic in there.

Might be worth looking into ways to improve this use case in v2 in the future. Any input here would be great. But kinda difficult due to limitations in DynamoDB.

Hopefully that answers your question. If not let me know and I can try to help further.

---

Tl;dr

> Or does DynamDB/Dynamoose just not operate as similarly to Mongo/Mongoose as I thought and I just need to go grab all the companies, and then filter them once they've been returned from the DB through logic on the server?

Yes. Does not operate as similarly to MongoDB/Mongoose in this instance (from what you've told me). Need to handle this on your own. And it is a limitation/difference in DynamoDB vs MongoDB, not Dynamoose vs Mongoose.

**Comment by z-murphydev** [comment]
@fishcharlie 

Excellent! Awesome response and thank you so much. Really went above and beyond here and I seriously appreciate it. So, I guess it makes more sense to think of Dynamo as closer to SQL. 

This is no longer directly related to dynamoose. So feel free to close this issue/mark as resolved, but if you don’t mind I’d like to pick your brain with one more question. 

So instead of storing an array associated with the id, it would make more sense to store every company as it’s own object, and each object is associated with that id.

As an example, let’s say I have 800 companies that I wan associated with a user. I would create 800 entries in dynamoose - where each entry has the userid in one column and a single company in the other. Would that work? Or can I not have multiple entries for a userid since it’s my...rangekey think is Dynamos term. So I’d need the adjust the schema to have a different unique identifier for each entry, yeah?

**Comment by fishcharlie** [comment]
@z-murphydev 

> So, I guess it makes more sense to think of Dynamo as closer to SQL.

I don't think DynamoDB being like SQL is a good comparison to make. I mean, it's possible with some hacks. But it's not the same at all, and if you do SQL like design at scale, it won't be very efficient. Highly recommend you check out this video from AWS re:Invent 2019: https://youtu.be/DIQVJqiSUkE. It goes over a lot of good design patterns for DynamoDB.

> So feel free to close this issue/mark as resolved, but if you don’t mind I’d like to pick your brain with one more question.

Actually starting a new policy where I'm trying to close less issues. Trying out a new bot that will close if an issue or PR has no reply for an extended period of time. So trying to let issues or PRs be closed as stale, or have them be closed by the author (which obviously is preferred).

> As an example, let’s say I have 800 companies that I wan associated with a user. I would create 800 entries in dynamoose - where each entry has the userid in one column and a single company in the other.

800 companies is a LOT. And the video I linked above doesn't really go into detail about that. You might find some that disagree with me about this, but the biggest thing I try to focus on is `get` and `query` in DynamoDB. I try to avoid `scan` at all costs. In your case with so much data, you are likely to run into the 400kb limit per document/item that DynamoDB has.

It all depends on your use case and what tradeoffs you wanna make. Just with the limited amount of information I have about your requirements and use case, yes, I'd probably do a system like join. Dynamoose is also coming out with populate (#858) which was a feature in v1 but bringing it back to v2 soon hopefully. But that feature is very similar to like a join. It stores the key to another table's document/item as an attribute, then you can call `document.populate` and it'll go and fetch that other document and set the document attribute to be that other document.

The worst thing about this system IMO, is that you are gonna have to make multiple requests. Because DynamoDB you can't join at the database level. So you will have to make multiple requests to the DB to get all the data you are looking for. Which honestly, really sucks. And probably the biggest reason to avoid join like things with DynamoDB.

But as I said, associating 800 items is a LOT. So populate might be useful here.

---

Ok final note. Highly encourage you to watch that video linked above. Again I don't know your specific use case, but it sounds semi similar to the video you are looking for. It might be a huge light bulb moment and might give you the vision and path forward you are looking for.

**Everything** in that video is possible to do with Dynamoose. However, I'm aware of some ways to make it more efficient and make the syntax better. Take a look at https://github.com/dynamoose/dynamoose/issues/768#issuecomment-608160537 with timestamps and actions I'm planning to take to improve Dynamoose to make that process better. Hoping to start work implementing those things ASAP, but gotta get through a few bugs that were found in v2 first.

Feel free to also join the Slack channel (link to join in README badge) if you prefer a more back and forth way of communicating. I'd be happy to help you further on this or anything else either through GitHub issues, or Slack, whichever you prefer!! 😃 

Hopefully this helps. Let me know if you have any additional questions and I'd be more than happy to help!

**Comment by Avery382** [comment]
@fishcharlie https://stackdev.space/questions/create-a-conditional-logic-for-multiplying-values-within-an-array-of-mongodb-queries
